src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs

```diff
@@ -2084,45 +2084,36 @@ private void VisitBinaryOperatorChildren(BoundBinaryOperator node)
             //
             // Of course we must ensure that we visit the left hand side before the right hand side.
             var stack = ArrayBuilder<BoundBinaryOperator>.GetInstance();
-            stack.Push(node);
 
-            BoundBinaryOperator binary;
-            BoundExpression child = node.Left;
-
-            while (true)
+            BoundBinaryOperator binary = node;
+            do
             {
-                binary = child as BoundBinaryOperator;
-                if (binary == null || binary.OperatorKind.IsLogical())
-                {
-                    break;
-                }
-
                 stack.Push(binary);
-                child = binary.Left;
+                binary = binary.Left as BoundBinaryOperator;
             }
+            while (binary != null && !binary.OperatorKind.IsLogical());
 
-            VisitRvalue(child);
+            VisitBinaryOperatorChildren(stack);
+            stack.Free();
+        }
+
+        protected virtual void VisitBinaryOperatorChildren(ArrayBuilder<BoundBinaryOperator> stack)
+        {
+            var binary = stack.Pop();
```

Comment:
If you're going to rearrange this, consider making this a do/while loop instead to avoid duplication.

Reply:
It would also make this clearer to read.

---
In reply to: [324294218](https://github.com/dotnet/roslyn/pull/38320#discussion_r324294218) [](ancestors = 324294218)

Reply:
I suspect a do-while might look like the following which looks complicated:
```
bool first = true;
do
{
    var binary = stack.Pop();
    if (first)
    {
        VisitRvalue(binary.Left);
        first = false;
    }
    VisitRvalue(binary.Right);
    if (stack.Count > 0)
    {
        Unsplit(); // VisitRvalue does this
    }
} while (stack.Count > 0);
```

---
In reply to: [324294303](https://github.com/dotnet/roslyn/pull/38320#discussion_r324294303) [](ancestors = 324294303,324294218)

Reply:
Alright, I wasn't imagining it would be that complicated, but think you're right.

---
In reply to: [324298721](https://github.com/dotnet/roslyn/pull/38320#discussion_r324298721) [](ancestors = 324298721,324294303,324294218)